rust/zypp-agama/src/callbacks/pkg_download.rs

```diff
@@ -0,0 +1,283 @@
+use std::os::raw::{c_char, c_void};
+
+use crate::{
+    callbacks::ProblemResponse,
+    helpers::{as_c_void, string_from_ptr},
+};
+
+pub enum DownloadError {
+    NoError,
+    NotFound, // the requested Url was not found
+    IO,       // IO error
+    Invalid,  // the downloaded file is invalid
+}
+
+impl From<zypp_agama_sys::DownloadResolvableError> for DownloadError {
+    fn from(error: zypp_agama_sys::DownloadResolvableError) -> Self {
+        match error {
+            zypp_agama_sys::DownloadResolvableError_DRE_NO_ERROR => DownloadError::NoError,
+            zypp_agama_sys::DownloadResolvableError_DRE_NOT_FOUND => DownloadError::NotFound,
+            zypp_agama_sys::DownloadResolvableError_DRE_IO => DownloadError::IO,
+            zypp_agama_sys::DownloadResolvableError_DRE_INVALID => DownloadError::Invalid,
+            _ => {
+                tracing::error!("Unknown error code {:?}", error);
+                DownloadError::Invalid
+            }
+        }
+    }
+}
+
+impl ToString for DownloadError {
+    fn to_string(&self) -> String {
+        match self {
+            DownloadError::NoError => "NoError".to_string(),
```

Comment:
NP: this works too
```diff
         match self {
-            DownloadError::NoError => "NoError".to_string(),
-            DownloadError::NotFound => "NotFound".to_string(),
-            DownloadError::IO => "IO".to_string(),
-            DownloadError::Invalid => "Invalid".to_string(),
+            DownloadError::NoError => "NoError",
+            DownloadError::NotFound => "NotFound",
+            DownloadError::IO => "IO",
+            DownloadError::Invalid => "Invalid",
         }
+        .to_string()
```

but also Clippy says that we should implement `Display` instead to get `ToString` for free, https://rust-lang.github.io/rust-clippy/rust-1.91.0/index.html#to_string_trait_impl